Menus/MainMenu/MainMenu.cs

```diff
@@ -143,77 +158,130 @@ void CreateFaceButtons(List<Type> types)
-					buttonData.sectionName = "Workspaces";
+					if (buttonData == null)
+						buttonData = new MainMenuUI.ButtonData { name = type.Name.Replace("Tool", string.Empty) };
+
+					CreateFaceButton(buttonData, tooltip, () =>
+					{
+						if (targetRayOrigin)
+						{
+							this.SelectTool(targetRayOrigin, selectedType);
+							UpdateToolButtons();
+						}
+					});
 				}
-				else if (isSettingsProvider)
+
+				if (isWorkspace)
 				{
 					// For workspaces that haven't specified a custom attribute, do some menu categorization automatically
-					buttonData.name = type.Name.Replace("Module", string.Empty);
-					buttonData.sectionName = "Settings";
+					if (buttonData == null)
+					{
+						buttonData = new MainMenuUI.ButtonData
+						{
+							name = type.Name.Replace("Workspace", string.Empty),
+							sectionName = "Workspaces"
+						};
+					}
+
+					CreateFaceButton(buttonData, tooltip, () =>
+					{
+						this.CreateWorkspace(selectedType);
+					});
 				}
 
-				var selectedType = type; // Local variable for proper closure
-				m_MainMenuUI.CreateFaceButton(buttonData, b =>
+				if (isSettingsProvider)
 				{
-					b.button.onClick.RemoveAllListeners();
-					if (isTool)
+					foreach (var providerPair in settingsMenuProviders)
 					{
-						m_ToolButtons[selectedType] = b;
-
-						b.button.onClick.AddListener(() =>
+						var kvp = providerPair.Key;
+						if (kvp.Key == type && (kvp.Value == null || kvp.Value == rayOrigin))
 						{
-							if (visible && targetRayOrigin)
+							var menuProvider = providerPair.Value;
+							if (buttonData == null)
 							{
-								this.SelectTool(targetRayOrigin, selectedType);
-								UpdateToolButtons();
+								buttonData = new MainMenuUI.ButtonData
+								{
+									name = type.Name.Replace("Tool", string.Empty).Replace("Module", string.Empty)
+								};
 							}
-						});
-					}
-					else if (isWorkspace)
-					{
-						b.button.onClick.AddListener(() =>
-						{
-							if (visible)
-								this.CreateWorkspace(selectedType);
-						});
+
+							buttonData.sectionName = "Settings";
+
+							CreateFaceButton(buttonData, tooltip, () =>
+							{
+								menuProvider.settingsMenuInstance = m_MainMenuUI.AddSubmenu(buttonData.sectionName, menuProvider.settingsMenuPrefab);
+							});
+						}
 					}
-					else if (isSettingsProvider)
+				}
+
+				if (isSettingsItemProvider)
+				{
+					foreach (var providerPair in settingsMenuItemProviders)
 					{
-						b.button.onClick.AddListener(() =>
+						var kvp = providerPair.Key;
+						if (kvp.Key == type && (kvp.Value == null || kvp.Value == rayOrigin))
 						{
-							var provider = settingsMenuProviders[selectedType];
-							provider.settingsMenuInstance = m_MainMenuUI.AddSubmenu(buttonData.sectionName, provider.settingsMenuPrefab);
-						});
+							var itemProvider = providerPair.Value;
+							if (buttonData == null)
+							{
+								buttonData = new MainMenuUI.ButtonData
```

Comment:
discussion result : doing name sanitizing.  If this ends up over-aggressive at some point we'll introduce a way to specify custom clean names